### PR TITLE
feature - mvvm register

### PR DESCRIPTION
**Pull request by at-bienle**
#### What's this PR do? Please give a technical summary `(required)`
- [x] register api
#### Any references (stackoverflow …) `(optional)`

### CODE REVIEW

File: MyApp/Model/API/Core/Api.swift [hunk 1]
```diff
@@ -10,10 +10,35 @@ import Foundation
...
+
+    }
+}
+
+extension Api.Path {
+    struct User: ApiPath {
+        static var path: String { return baseURL / "v1/user" }
+        var urlString: String { return User.path / "register" }
     }
 }
```
Comment (at-daiho) [inline MyApp/Model/API/Core/Api.swift]: WARNING - There should be a space after //

**Comment by at-daiho** [comment]
Total linter violations in pull request: 1
Serious: 0
Warnings: 1

**Comment by at-daiho** [comment]
:+1: There are no linter violations.

File: MyApp/Model/API/Api.Register.swift [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        let authId: String
+        let authToken: String
+        let authMethod = "email"
+    }
+
+    // MARK: - Public
+    @discardableResult
+    static func register(params: RegisterParams, completion: @escaping Completion) -> Request? {
+        let path = Api.Path.Register()
+        let param: Parameters = [
```
Comment (at-linhvo) [inline MyApp/Model/API/Api.Register.swift]: write: func toJson() in Struct RegisterParams

File: MyApp/Model/API/Core/RequestSerializer.swift [hunk 3]
```diff
@@ -13,7 +13,7 @@ extension ApiManager {
     @discardableResult
     func request(method: HTTPMethod,
                  urlString: URLStringConvertible,
-                 parameters: [String: Any]? = nil,
+                 parameters: Parameters? = nil,
```
Comment (at-linhvo) [inline MyApp/Model/API/Core/RequestSerializer.swift]: don't allow edit in class RequestSerializer

File: MyApp/View/Controllers/Register/RegisterViewController.swift [hunk 4]
```diff
@@ -27,8 +29,43 @@ class RegisterViewController: ViewController {
...
     @IBAction func resetButtonTouchUpInside(_ sender: UIButton) {
+        userNameTextField.text = ""
+        passwordTextField.text = ""
+        emailTextField.text = ""
+        confirmPasswordTextField.text = ""
+    }
+}
+
+// MARK: - Extension RegisterViewController
+extension RegisterViewController {
```
Comment (at-linhvo) [inline MyApp/View/Controllers/Register/RegisterViewController.swift]: MARK: - Function in RegisterVC. No extension.

File: MyApp/ViewModel/RegisterViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    // MARK: - Public
+    func register(_ completion: @escaping (RegisterResult) -> Void) {
+        let validation = validate()
+        guard validation.isSuccess else {
+            let info: [String: Any] = [
+                NSLocalizedDescriptionKey: validation.description
+            ]
+            let error = NSError(domain: "", code: -1, userInfo: info)
```
Comment (at-linhvo) [inline MyApp/ViewModel/RegisterViewModel.swift]: do you understand this code?

**Pull request merged**